### PR TITLE
Fallback cause is propagated to fallback provider

### PR DESCRIPTION
**Pull request by DominikMostek**
Cause can now be propagated to fallback provider in order to respond accordingly to a exception type.
New interface is introduced in order not to break any existing falback providers.

### CODE REVIEW

**Comment by DominikMostek** [comment]
Hi, this is my first contribution, hope it will be appreciated. In our project we would like to return different responses and log some details for different exceptions. Looking forward to your feedback and comments.

**Comment by spencergibb** [comment]
If we're going to deprecate the old, I'd like a better name for the new one.

**Comment by DominikMostek** [comment]
@ryanjbaxter No problem with creating a cleanup PR 

@spencergibb Any suggestions for a better name? I also don't like this one very much, but was not able to come up with better

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/ZuulFallbackProviderWithCause.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.springframework.http.client.ClientHttpResponse;
+
+/**
+ * Extension of {@link ZuulFallbackProvider} which adds possibility to choose proper response
+ * based on the exception which caused the main method to fail.
+ *
+ * @author Dominik Mostek
+ */
+public interface ZuulFallbackProviderWithCause extends ZuulFallbackProvider {
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/ZuulFallbackProviderWithCause.java]: Maybe something like `ThrowableZuulFallbackProvider`?

Comment (spencergibb) [reply spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/ZuulFallbackProviderWithCause.java]: FallbackProvider?

Comment (ryanjbaxter) [reply spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/ZuulFallbackProviderWithCause.java]: That works.

**Comment by DominikMostek** [comment]
I've updated PR with new name and documentation. Also created a [2.0.x based branch with deprecation]( https://github.com/DominikMostek/spring-cloud-netflix/tree/fallback_provider_2.0) cleanup (it has cherry-picked commits from this PR) 

The tests are failing now, but honestly I am not sure why.

**Comment by ryanjbaxter** [comment]
The build is just being flakey.  Could you submit a PR for the changes in the 2.0.x branch?

**Comment by spencergibb** [comment]
Won't we get them in a merge forward?

**Comment by ryanjbaxter** [comment]
Yes, I just wanted to changes to remove the old deprecated class and fix the documentation.  I suppose @DominikMostek will have to wait for us to merge this and then merge it forward before doing that though.

**Comment by spencergibb** [comment]
Oh right, yeah he'll have to wait until merged forward.

**Comment by ryanjbaxter** [comment]
@DominikMostek these changes are now in the 2.0.x branch as well.  You can go ahead and remove the deprecated class and cleanup the docs in your 2.0.x branch and submit a PR.  Thanks!

**Pull request merged**